### PR TITLE
Added NXP SE05x support (secure element)

### PR DESCRIPTION
**Pull request by AndreHeinemans-NXP**
## Summary
Added NXP SE05x support (secure element)
Depends on tooling in [Apps (PR pending)](https://github.com/apache/nuttx-apps/pull/1975) 

## Impact
ucans32k146:se05x target has this driver enabled

## Testing

please ignore mix case error:
```
Error: /home/runner/work/nuttx/nuttx/nuttx/drivers/crypto/pnt/pnt_util.h:81:20: error: Mixed case identifier found
```

### CODE REVIEW

File: drivers/crypto/pnt/Make.defs
Comment (xiaoxiang781216) [inline drivers/crypto/pnt/Make.defs:79]: remove vpath for crypto

File: include/nuttx/crypto/se05x.h
Comment (xiaoxiang781216) [inline include/nuttx/crypto/se05x.h:76]:
```suggestion
} se05x_algorithm_e;
```

File: include/nuttx/crypto/se05x.h
Comment (xiaoxiang781216) [inline include/nuttx/crypto/se05x.h:64]:
```suggestion
} se05x_asym_cipher_type_e;
```

File: drivers/crypto/se05x_internal.h
Comment (xiaoxiang781216) [inline drivers/crypto/se05x_internal.h:46]:
```suggestion
  mutex_t mutex;
```
and call mutex_xxx api

File: drivers/crypto/se05x_internal.h
Comment (xiaoxiang781216) [inline drivers/crypto/se05x_internal.h:2]: why not move into pnt folder

File: drivers/crypto/se05x.c
Comment (xiaoxiang781216) [inline drivers/crypto/se05x.c:2]: why not move into pnt folder

**Review by acassis — COMMENTED** [review]
Thank you @AndreHeinemans-NXP !
Please follow the NuttX Coding Style: https://nuttx.apache.org/docs/latest/contributing/coding_style.html

You can also run the checkpatch to verify your files:
$ ./tools/checkpatch.sh -f path/to/your/file.c

File: drivers/crypto/se05x.c
Comment (AndreHeinemans-NXP) [reply drivers/crypto/se05x.c:2]: The pnt folder only contains the external 'plug&trust' library and the needed wrappers. The intention was to seperate that from the ioctl functions

**Comment by raiden00pl** [comment]
I'm a little concerned about this change.

It seems to me that NuttX has always avoided the approach where we pull code from the
ventor's repositories and wrap it in some code to work with RTOS.

This approach silently evades the NuttX coding standard, make code review more difficult
and make it impossible to reuse code in similar crypto drivers.

It think for drivers only drivers/segger has a similar approach, but in this case it's justified,
as the driver is dedicated to the Segger tools.

I understand that this approach is easier for NXP, but from the project point of view,
it seems wrong.

If it's not necessary because of some precompiled files (closed source drivers for chip peripherals)
or a solution commonly considered a standard in the embedded world (jlink tools, openamp),
we shouldn't do that.

Please, let's not make NuttX another Zephyr where vendors take shortcuts and instead of
creating code dedicated for the OS, they download a ton of code from their repositories.

**Comment by raiden00pl** [comment]
BTW do we have CCLA from NXP ? I don't see in `AUTHORS`. When does it become required?

**Comment by PetervdPerk-NXP** [comment]
We try to write a dedicated code for NuttX whenever possible i.e. S32K1, S32K3, IMXRT and SocketCAN.
For the SE050 secure element we took a different approach because of complexity of the Application Protocol Data Unit (APDU) and maintaining the security updates through the external repository. The [repository ](https://github.com/NXPPlugNTrust/nano-package)we use is open-source and Apache 2.0 licensed.

> make it impossible to reuse code in similar crypto drivers.

We took an approach to abstract the Secure Element through the introduced ` _SEIOCBASE` ioctl. With generic commands such as `SEIOC_GENERATE_KEYPAIR` any other crypto/secure element driver can reuse this interface.

Also how come it's okay for a wireless driver i.e. ESP32, ESP32C3, BL602, NRF52, NRF53 and NRF91 to pull from external repositories and now for a secure element it's not?

**Comment by raiden00pl** [comment]
> We try to write a dedicated code for NuttX whenever possible i.e. S32K1, S32K3, IMXRT and SocketCAN.

Thanks for that :)

> For the SE050 secure element we took a different approach because of complexity of the Application Protocol Data Unit (APDU) and maintaining the security updates through the external repository.

That's another thing. Is this a good approach to rely on security updates in an external repo? Teams in companies change, projects are abandoned, and the response time of corporations to security flaws often leaves much to be desired.

> Also how come it's okay for a wireless driver i.e. ESP32, ESP32C3, BL602, NRF52, NRF53 and NRF91 to pull from external repositories and now for a secure element it's not?

In these cases, the downloaded code is an interface to the binary files provided by the vendor.
You get a precompiled, certified wireless stack. It can't be done otherwise unless we write wireless stack from scratch - but that requires open radio peripheral documentation (true for Nordic, not true for ESP) and resources which NuttX community doesn't have.

At least for Nordic chips the amount of pulled code is minimal to make it work. Not sure about the rest.
For NRF91 there is even an IPC implementation dedicated to NuttX that is functionally compatible with
the Nordic SDK implementation, just to avoid dependence on the vendor's code (https://github.com/apache/nuttx/blob/master/arch/arm/src/nrf91/nrf91_nrfx_ipc.c).

**Comment by PetervdPerk-NXP** [comment]
> In these cases, the downloaded code is an interface to the binary files provided by the vendor.
You get a precompiled, certified wireless stack. It can't be done otherwise unless we write wireless stack from scratch - but that requires open radio peripheral documentation (true for Nordic, not true for ESP) and resources which NuttX community doesn't have.

But isn't that what were doing here? You get a vendor provided secure element stack. It can't be done otherwise unless we write secure element stack from scratch. But that requires open security manual (false for NXP).

**Comment by raiden00pl** [comment]
> But isn't that what were doing here? You get a vendor provided secure element stack. It can't be done otherwise unless we write secure element stack from scratch. But that requires open security manual (false for NXP).

If APDU uses any precompiled binary files provided by NXP or APDU interface is not openly available, then I don't see any other option than downloading from the vendor. 
If both statements are false, then getting the code from the vendor is a shortcut. That's how I see it.

The question remains what it means that the interface is open. No NDA? No login barrier? Any available source from which we can get interface definitions? This question is more for the wider community.

Zephyr-ation (duplicating Zephyr's bad practices which are convenient for vendors) is the worst thing that can happen to this project, so that's why I'm complaining :)

**Comment by raiden00pl** [comment]
> So if we hide it in a blob it's okay, but if there's a source available it's called zephyr-ation?
We could just publish a blob and use that, but would be bad practice in my opinion.

If you hide it in a blob AND you don't provide open documentation for that chip then yes, I think it's OK. Or if you provide some kind of certification for a given solution. In other cases, precompiled binary is the worst possible option.

>  Pretty much how LittleFS is handled in NuttX as well

littlefs is a well-established open source project independent of any chip vendor (just like previously mentioned openamp). The chance that the project will be completely abandoned overnight and thrown into the abyss is minimal.

> If you find this kind of code-reuse a shortcut that would be your call but I think it's more effort then publishing a blob and make bindings to it.

I'm not going to block this PR in any way, it's just my dislike of this approach. If there are no more people with doubts then I don't have a problem with that. But if more similar solutions start to come out, or worse, we'll start relying on the vendors HAL then I think it's going to be a problem.

**Comment by jarivanewijk** [comment]
> BTW do we have CCLA from NXP ? I don't see in `AUTHORS`. When does it become required?

I don't want to disrupt the discussion on this PR, but I got triggered by your comment. To my knowledge a CCLA has been submitted by NXP over a year ago. Some colleagues (including myself) have also submitted ICLAs. I must have confirmation somewhere that it was accepted by the Apache Foundation. Looks like this still needs to be added to the AUTHORS file? Should we have made a PR for that ourselves? I would have expected that file to get updated periodically.

**Comment by raiden00pl** [comment]
> I don't want to disrupt the discussion on this PR, but I got triggered by your comment. To my knowledge a CCLA has been submitted by NXP over a year ago. Some colleagues (including myself) have also submitted ICLAs. I must have confirmation somewhere that it was accepted by the Apache Foundation. Looks like this still needs to be added to the AUTHORS file? Should we have made a PR for that ourselves? I would have expected that file to get updated periodically.

It'll be great when AUTHORS gets updated. This will help to avoid doubts like mine in the future.

**Comment by AndreHeinemans-NXP** [comment]
> @AndreHeinemans-NXP please rebase your patch to the last master.

Done

**Comment by xiaoxiang781216** [comment]
@AndreHeinemans-NXP you have to fix this warning:
```
In file included from mbedtls/library/ssl_misc.h:27,
                 from mbedtls/library/constant_time_internal.h:30,
                 from mbedtls/library/base64.c:25:
Error: /github/workspace/sources/apps/crypto/mbedtls/mbedtls/include/mbedtls/ssl.h:590:5: error: "MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT" is not defined, evaluates to 0 [-Werror=undef]
  590 | #if MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT == 0
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
before merging this patch.

**Comment by AndreHeinemans-NXP** [comment]
@xiaoxiang781216
This problem cannot be fixed in this changeset because it was already there in NuttX-apps/MbedTLS. If you build a random target on master with MbedTLS enabled then these warnings will appear.

I think the problem is in MbedTLS' ssl.h:590 which should also test MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT for not being defined at all. But that file is downloaded and not part of the NuttX or NuttX-apps repositories.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 This problem cannot be fixed in this changeset because it was already there in NuttX-apps/MbedTLS. If you build a random target on master with MbedTLS enabled then these warnings will appear.
> 
> I think the problem is in MbedTLS' ssl.h:590 which should also test MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT for not being defined at all. But that file is downloaded and not part of the NuttX or NuttX-apps repositories.

it is fixed here: https://github.com/apache/nuttx-apps/pull/1995

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
@xiaoxiang781216 is this ready to be merged?

**Comment by xiaoxiang781216** [comment]
arm-05 ci fail, could you rebase the change to the last master?

**Comment by xiaoxiang781216** [comment]
@PetervdPerk-NXP the follow compiler error need be fixed first:
```
In file included from mbedtls/library/ssl_misc.h:27,
                 from mbedtls/library/constant_time_internal.h:30,
                 from mbedtls/library/base64.c:25:
Error: /github/workspace/sources/apps/crypto/mbedtls/mbedtls/include/mbedtls/ssl.h:590:5: error: "MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT" is not defined, evaluates to 0 [-Werror=undef]
  590 | #if MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT == 0
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by AndreHeinemans-NXP** [comment]
@xiaoxiang781216 

The problem is in MbedTLS like I mentioned earlier but I found a workaround by also enabling MBEDTLS_SSL_PROTO_DTLS and MBEDTLS_SSL_DTLS_CONNECTION_ID in the config. These changes are pushed which should fix the build

I added this issue in apache/nuttx-apps#2045

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> 
> The problem is in MbedTLS like I mentioned earlier but I found a workaround by also enabling MBEDTLS_SSL_PROTO_DTLS and MBEDTLS_SSL_DTLS_CONNECTION_ID in the config. These changes are pushed which should fix the build
> 
> I added this issue in [apache/nuttx-apps#2045](https://github.com/apache/nuttx-apps/issues/2045)

could you provide  a patch to fix this issue?

**Comment by AndreHeinemans-NXP** [comment]
@xiaoxiang781216 

```
diff --git a/boards/arm/s32k1xx/ucans32k146/configs/se05x/defconfig b/boards/arm/s32k1xx/ucans32k146/configs/se05x/defconfig
index a6ba699d9b..a7bafc199c 100644
--- a/boards/arm/s32k1xx/ucans32k146/configs/se05x/defconfig
+++ b/boards/arm/s32k1xx/ucans32k146/configs/se05x/defconfig
@@ -32,6 +32,8 @@ CONFIG_HAVE_CXXINITIALIZE=y
 CONFIG_I2C=y
 CONFIG_INIT_ENTRYPOINT="nsh_main"
 CONFIG_LPUART1_SERIAL_CONSOLE=y
+CONFIG_MBEDTLS_SSL_DTLS_CONNECTION_ID=y
+CONFIG_MBEDTLS_SSL_PROTO_DTLS=y
 CONFIG_NSH_ARCHINIT=y
 CONFIG_NSH_BUILTIN_APPS=y
 CONFIG_NSH_FILEIOSIZE=512
```

its already pushed

**Pull request merged**
